KAFKA-8526: logdir fallback on getOrCreateLog

soarez opened the pull request:
LogManager#getOrCreateLog() selects a log dir for the new replica from
 _liveLogDirs, if disk failure is discovered at this point, before
 LogDirFailureHandler finds out, try using other log dirs before failing
 the operation.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

soarez committed:
KAFKA-8526: logdir fallback on getOrCreateLog

 LogManager#getOrCreateLog() selects a log dir for the new replica from
 _liveLogDirs, if disk failure is discovered at this point, before
 LogDirFailureHandler finds out, try using other log dirs before failing
 the operation.

soarez commented:
pinging @apovzner ; can we get some feedback on this PR?

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -691,48 +692,67 @@ class LogManager(logDirs: Seq[File],
+          .recoverWith {
+            case e: KafkaStorageException =>
+              // If the next log directory, try the other ones before failing
+              _liveLogDirs.asScala
```
I have a minor suggestion. This should be a rare case, so it may not matter too much, but could we let `nextLogDir` return the full ordered list of suggested directories rather than just a single entry? Then we would still take load into account when retrying. It would also let us move this loop to the top level so that `createLogDirectory` is only called in one spot.

Something roughly like this:
```scala
val logDirs = {

  // omitted

  if (preferredLogDir != null)
    List(preferredLogDir)
  else
    nextLogDirs()
}

val dir = logDirs
  .find(createLogDirectory(_, logDirName).isSuccess)
  .get
```

soarez replied:
Thanks for the review @hachikuji. I like the suggestion, makes things simpler. Added this in b6c36f1. Please take another look.

soarez committed:
nextLogDirs returns ordered list of suggested dirs

nextLogDir now returns the full ordered list of suggested directories rather
than just a single entry.

The list provided by nextLogDirs takes load into account.

This list is then scanned for the first entry where we can successfully
create a log directory.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -687,55 +688,69 @@ class LogManager(logDirs: Seq[File],
+        val logDir = logDirs
+          .map(createLogDirectory(_, logDirName))
+          .find(_.isSuccess)
+          .getOrElse(Failure(new KafkaStorageException("No log directories available")))
```
Perhaps this message can indicate which log directories were tried?

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -687,55 +688,69 @@ class LogManager(logDirs: Seq[File],
+        case e: IOException =>
+          val msg = s"Error while creating log for $logDirName in dir $logDirPath"
+          logDirFailureChannel.maybeAddOfflineLogDir(logDirPath, msg, e)
+          warn(e.getMessage, e)
```
Maybe we can just use `msg` instead of `e.getMessage`? I think we'll get the exception message from the log output anyway.

hachikuji commented on core/src/test/scala/unit/kafka/log/LogManagerTest.scala:
```diff
@@ -95,6 +96,33 @@ class LogManagerTest {
+    logManager.startup()
+
+    // To simulate disk failure, delete log directories and replace them with files
+    // so that Files.createDirectories will fail.
```
I guess another way to do this might be extend `Log` and override `createLogDirectory`. Also, to remove the need for 100 directories, perhaps we could make the sorting more deterministic. For example, maybe we can sort by log count first and name second?

soarez replied:
That's a good idea. Even with the 100 directories, with some weird sorting it could still fail.
I'd like to avoid adding further sorting work just for the sake of testing. I think I've got a way to do it reliably without extra sorting. Looking forward to your thoughts on it.

hachikuji reviewed (commented):
Thanks, looks good overall. Left a couple additional comments.

hachikuji commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -687,55 +688,69 @@ class LogManager(logDirs: Seq[File],
-          // Remove the preferred log dir since it has already been satisfied
-          preferredLogDirs.remove(topicPartition)
+        val logDir = logDirs
+          .map(createLogDirectory(_, logDirName))
```
Unfortunately, the `map` here doesn't work because it will create the partition directory for all directories even when there are no failures. I think we only want to call `createLogDirectory` on a new directory if the previous one failed.

soarez replied:
That's right, forgot a `.toStream` there.

soarez replied:
That's right, forgot a `.toStream` there.

apovzner replied:
I agree that we only want to call `createLogDirectory` if the previous one failed.

apovzner replied:
Oh this code actually looks right, nm. Since we call find(), it will stop on the first success.

soarez committed:
Prevent createLogDirectory being called for every directory

soarez committed:
Use msg instead of e.getMessage

soarez committed:
Indicate which log directories were tried

soarez committed:
Extend Log and override createLogDirectory for testing

soarez commented:
Thanks for the excellent review @hachikuji .
I believe I've addressed your points, please take another look.

hachikuji commented on core/src/test/scala/unit/kafka/utils/TestUtils.scala:
```diff
@@ -1012,13 +1013,18 @@ object TestUtils extends Logging {
                        cleanerConfig: CleanerConfig = CleanerConfig(enableCleaner = false),
-                       time: MockTime = new MockTime()): LogManager = {
+                       time: MockTime = new MockTime(),
+                       createLogDirectoryOverride: CreateLogDirectoryOverride = original => original
```
Since this is really a one-off need, it would be nice not to clutter up `TestUtils`. Would it be possible to construct the regular `LogManager` object in `testCreateLogWithLogDirFallback` and then use a Mockito spy to alter the behavior of `createLogDirectory`?

soarez replied:
That makes sense. I think I got that in d9742cb, please take another look.

hachikuji reviewed (commented):
Thanks, the fix is looking good. Left one comment on the test case.

soarez committed:
Revert "Extend Log and override createLogDirectory for testing"

This reverts commit bc629c8a018ba5cd60f287fd294c1e1c4a123752.

soarez committed:
Use a spy to alter behavior of createLogDirectory

hachikuji reviewed (approved):
LGTM. Thanks for the patch!

hachikuji commented:
retest this please

hachikuji commented:
retest this please

hachikuji commented:
retest this please

hachikuji commented:
retest this please

hachikuji committed:
Fix compilation error on scala 2.11

apovzner reviewed (approved):
This is great! I also wrote a small unit test to verify KAFKA-8526 before I opened the JIRA. Instead of mocking, I just made one dir read only. I verified that my unit test passed with this PR. Thanks for the patch!

hachikuji commented:
@apovzner Thanks for for the additional review. I will try one more time to get a clean build and then merge.

hachikuji commented:
retest this please

hachikuji commented:
Going to go ahead and merge. The one failing test is a new addition which seems to be flaky. We can address this separately.

Thanks for the patch!
